File: var/lib/asterisk/agi-bin/yenimusterikaydi.php
#!/usr/bin/php
<?php

require 'phpagi.php';
//require '/var/lib/asterisk/agi-bin/vendor/autoload.php';
//use Aws\S3\S3Client;
//use Aws\TranscribeService\TranscribeServiceClient;
$agi = new AGI();
try {
    $agi->answer();
	$recordId= uniqid();
    // Ses kaydını başlat
    $record_file = '/var/spool/asterisk/monitor/temp_input_'.$recordId.'';
   
   // $agi->exec('Playback', 'silence/1');
    //$agi->record_file($record_file, 'wav', '#', -1, 0, false, 0);
    //$agi->record_file($record_file, 'wav', '#',3000, 3, true, 2000);
   // $agi->record_file($record_file, 'wav', '#', 10000, 5000, false, 0);
    //$agi->record_file($record_file, 'wav', '', -1, 0, false, 2);
    $agi->record_file($record_file, "wav", "", 3000, 0, false, 2);
    // Kaydedilen dosyayı STT API'ye gönder
    //$stt_result = transcribe_audio($record_file . '.wav');

    // Gelen metni analiz et
    //$destination = determine_destination($stt_result);

    // Yönlendirme
   

// Ses dosyasının yolu
$recordedFilePath = '/var/spool/asterisk/monitor/temp_input_'.$recordId.'.wav';
$command = "node /var/lib/asterisk/agi-bin/transcribe2.js " . escapeshellarg($recordedFilePath);
$output = shell_exec($command);

// JSON çıktısını çözümleyin
$result = json_decode($output, true);

if ($result['success']) {
	$id = uniqid();

	$agi->verbose("Transcription: " . $result['transcription'] . "\n");
	$yenimusteri = yenimusteri($result['transcription'],$argv[1],$agi);
	$agi->verbose('Başarılı : '.$yenimusteri['success']);
		$agi->set_variable('AGISTATUS',$yenimusteri['success']? "success":"failure");	
		$agi->set_variable('yeniMusteriHosgeldin',$yenimusteri['message']);
	        $agi->set_variable('userId',$yenimusteri["userId"]);
	
} else {
	$agi->verbose("transcription Error: " . $result['error'] . "\n");
}


} catch (Exception $e) {
    $agi->verbose('Hata: ' . $e->getMessage());
    $agi->hangup();
}
function yenimusteri($transcribe,$salonid,$agi)
{
    $url = 'https://app.randevumcepte.com.tr/api/v1/yenimusteridanisankaydi';  // Laravel API URL'si
    $data = ['santraldenkayit'=>true,'name' => $transcribe,'salonidler'=>$salonid,'cep_telefon'=>$agi->request["agi_callerid"]];  // Gönderilecek veri
    $agi->verbose('Müşteri adı : '.$transcribe." salon id : ".$salonid);
    // cURL başlatma
    $ch = curl_init($url);

    // cURL seçeneklerini ayarlama
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));  // JSON formatında veri gönderme
    curl_setopt($ch, CURLOPT_HTTPHEADER, [
        'Content-Type: application/json',  // İçerik tipi belirtme
    ]);

    // API yanıtını al
    $response = curl_exec($ch);

    // cURL hatasını kontrol et
    if (curl_errno($ch)) {
        $error_msg = curl_error($ch);
        error_log('cURL Hatası: ' . $error_msg);
	$agi->verbose('cURL Hatası: ' . $error_msg);

    } else {
        $agi->verbose('cURL yanıtı alındı.');
    }

    // cURL bağlantısını kapatma
    
    curl_close($ch);
    
    // Yanıtı JSON formatında çözümle
    $decoded_response = json_decode($response, true);

    if (json_last_error() !== JSON_ERROR_NONE) {
        $json_error = json_last_error_msg();
	error_log('JSON çözümleme hatası: ' . $json_error);
	$agi->verbose($response);
	$agi->verbose('JSON çözümleme hatası: ' . $json_error);
	error_log("Gelen JSON:\n" . json_encode(json_decode($response), JSON_PRETTY_PRINT));
    } else {
        $agi->verbose('JSON çözümleme başarılı. Gelen json '.json_encode(json_decode($response)));
    }

    return $decoded_response;
}



?>

File: var/lib/asterisk/agi-bin/yeniMusteriKaydi.py
#!/usr/bin/env python3
import sys
import os
import subprocess
import json
import uuid

# Minimal AGI sınıfı
class AGI:
	def __init__(self):
		self.vars = {}
		self.env = {}
		self.read_agi_env()

	def read_agi_env(self):
		while True:
			line = sys.stdin.readline().strip()
			if line == '':
				break
			if ':' in line:
				key, value = line.split(':', 1)
				self.env[key.strip()] = value.strip()

	def answer(self):
		print("ANSWER")
		sys.stdout.flush()

	def hangup(self):
		print("HANGUP")
		sys.stdout.flush()

	def verbose(self, msg):
		print(f'VERBOSE "{msg}" 1')
		sys.stdout.flush()

	def record_file(self, file_path, format="wav", escape_digits="", max_seconds=30, beep=False, silence=2):
		# AGI RECORD FILE komutunu elle gönderiyoruz
		beep_flag = 1 if beep else 0
		cmd = f'RECORD FILE {file_path} {format} "{escape_digits}" {max_seconds} {beep_flag} {silence}'
		print(cmd)
		sys.stdout.flush()
		# Asterisk'ten yanıtı oku
		sys.stdin.readline()

	def set_variable(self, name, value):
		print(f'SET VARIABLE {name} "{value}"')
		sys.stdout.flush()

# AGI başlat
agi = AGI()

try:
	agi.answer()

	# Benzersiz kayıt dosyası
	record_id = str(uuid.uuid4())
	recordedFilePath = f"/var/spool/asterisk/monitor/temp_input_{record_id}"

	agi.verbose(f"Recording to {recordedFilePath}")

	# 30 saniye kayıt, bip kapalı, 2 saniye sessizlik ile durdur
	agi.record_file(recordedFilePath, "wav", "", 30, beep=False, silence=2)

	# Node.js transcribe çağrısı
	command = ["/usr/bin/node", "/var/lib/asterisk/agi-bin/transcribe2.js", recordedFilePath]

	try:
		output = subprocess.check_output(command, stderr=subprocess.STDOUT).decode()
	except subprocess.CalledProcessError as e:
		output = e.output.decode()

	agi.verbose(f"Node output: {output}")

	try:
		result = json.loads(output)
	except json.JSONDecodeError as e:
		agi.verbose(f"JSON decode error: {e}")
		result = {"success": False, "error": str(e)}

	if result.get("success"):
		agi.verbose(f"Transcription: {result.get('transcription')}")
		agi.set_variable("AGISTATUS", "success")
		agi.set_variable("TRANSCRIPT", result.get("transcription"))
	else:
		agi.verbose(f"Transcription failed: {result.get('error')}")
		agi.set_variable("AGISTATUS", "failure")

except Exception as e:
	agi.verbose(f"AGI Exception: {e}")
	agi.hangup()
